{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "9li5Jzka3KH7"
   },
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install SeqIO\n",
    "!pip install Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "QwLqbMuA3vn4"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "LuBP1KIMUIH6"
   },
   "source": [
    "# Reading and sorting dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -- Function for reading Dfam dataset in csv format\n",
    "def read_dataset(input_path):\n",
    "    df = pd.read_csv(input_path, low_memory = False)   # Low_memory argument for reading whole file\n",
    "    return df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "VZ9dXlXtJjAT",
    "outputId": "db536325-4495-4dc8-c489-f7ef264de99b"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100000"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -- Read dataset (locally) \n",
    "path = \"/Users/leandrojorqueravalero/Desktop/PhD/synthetic_ORF2/data/preprocessed_LINE_v2.csv\"\n",
    "df=read_dataset(path)\n",
    "len(df.index)   # Check complete reading "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "_JdIihZ_sW_S",
    "outputId": "6cbd430a-ebc8-4234-dff2-217112cd17e8"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26431"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -- Select desired columns and rename\n",
    "\n",
    "sorted_df = df.iloc[:,[19, 3,8, 18, 16]]   # 3: length  // 8:DNA_seq // 16: species data // 18: counts info // 19:ID + Dfam entry\n",
    "sorted_df.columns = ['ID','Length','DNA_seq', 'Counts', 'Species']\n",
    "#sorted_df.head(5)\n",
    "\n",
    "# -- Select columns matching required criteria (length)\n",
    "filtered_df = sorted_df.loc[sorted_df['Length']>3000]\n",
    "#len(filtered_df.index)\n",
    "\n",
    "# -- removing duplicates\n",
    "filtered_df.drop_duplicates(subset=['ID'], keep=False)\n",
    "len(filtered_df.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieving species name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1M5_orf2DF0000008</td>\n",
       "      <td>Eutheria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1M1_5endDF0000227</td>\n",
       "      <td>Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1M1_orf2DF0000228</td>\n",
       "      <td>Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1M2_orf2DF0000236</td>\n",
       "      <td>Eutheria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1M3_orf2DF0000244</td>\n",
       "      <td>Eutheria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1353</th>\n",
       "      <td>TE3262_SO2_FAM455DF0289703</td>\n",
       "      <td>Sitophilus oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1354</th>\n",
       "      <td>TE3263_SO2_FAM455DF0289704</td>\n",
       "      <td>Sitophilus oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>TE3264_SO2_FAM465DF0289705</td>\n",
       "      <td>Sitophilus oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1356</th>\n",
       "      <td>TE3265_SO2_FAM465DF0289706</td>\n",
       "      <td>Sitophilus oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1357</th>\n",
       "      <td>TE3269_SO2_FAM497DF0289710</td>\n",
       "      <td>Sitophilus oryzae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1358 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ID           Taxonomy\n",
       "0             L1M5_orf2DF0000008           Eutheria\n",
       "1             L1M1_5endDF0000227           Primates\n",
       "2             L1M1_orf2DF0000228           Primates\n",
       "3             L1M2_orf2DF0000236           Eutheria\n",
       "4             L1M3_orf2DF0000244           Eutheria\n",
       "...                          ...                ...\n",
       "1353  TE3262_SO2_FAM455DF0289703  Sitophilus oryzae\n",
       "1354  TE3263_SO2_FAM455DF0289704  Sitophilus oryzae\n",
       "1355  TE3264_SO2_FAM465DF0289705  Sitophilus oryzae\n",
       "1356  TE3265_SO2_FAM465DF0289706  Sitophilus oryzae\n",
       "1357  TE3269_SO2_FAM497DF0289710  Sitophilus oryzae\n",
       "\n",
       "[1358 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy = filtered_df['Species']\n",
    "name = filtered_df['ID'] \n",
    "species_dict = {}\n",
    "for i in range(len(filtered_df)):     # len(filtered_df)\n",
    "    ID = name.iloc[i]\n",
    "    species = taxonomy.iloc[i]\n",
    "    #print(ID, species)\n",
    "    split = species.split(\";\")\n",
    "    #print(split)\n",
    "    spec_name = split[-1]\n",
    "    clean_name = spec_name.strip(\"']\")\n",
    "    species_dict[ID] = clean_name\n",
    "\n",
    "# -- Show ouput dict\n",
    "species_dict\n",
    "\n",
    "# Make a df and ID column is set as index\n",
    "taxo_df = pd.DataFrame.from_dict(species_dict, orient='index') \n",
    "L1_taxo = taxo_df.reset_index() \n",
    "L1_taxo.columns = ['ID', 'Taxonomy']\n",
    "L1_taxo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "eukarya_classes = ['Mammalia', 'Aves', 'Reptilia', 'Actinopterygii', 'Amphibia', 'Insecta', 'Fungi', 'Plantae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1P1_orf2DF0000316</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1M5_orf2DF0000008</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1M1_5endDF0000227</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1M1_orf2DF0000228</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1M2_orf2DF0000236</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1353</th>\n",
       "      <td>RTE1DF0001521</td>\n",
       "      <td>non_listed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1354</th>\n",
       "      <td>LINE2C1_CEDF0004135</td>\n",
       "      <td>non_listed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>L1-1_TestuDF0004210</td>\n",
       "      <td>non_listed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1356</th>\n",
       "      <td>RTE-1_TestuDF0004211</td>\n",
       "      <td>non_listed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1357</th>\n",
       "      <td>RTE-2_TestuDF0004212</td>\n",
       "      <td>non_listed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1358 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        ID    Taxonomy\n",
       "0       L1P1_orf2DF0000316    Mammalia\n",
       "1       L1M5_orf2DF0000008    Mammalia\n",
       "2       L1M1_5endDF0000227    Mammalia\n",
       "3       L1M1_orf2DF0000228    Mammalia\n",
       "4       L1M2_orf2DF0000236    Mammalia\n",
       "...                    ...         ...\n",
       "1353         RTE1DF0001521  non_listed\n",
       "1354   LINE2C1_CEDF0004135  non_listed\n",
       "1355   L1-1_TestuDF0004210  non_listed\n",
       "1356  RTE-1_TestuDF0004211  non_listed\n",
       "1357  RTE-2_TestuDF0004212  non_listed\n",
       "\n",
       "[1358 rows x 2 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy = filtered_df['Species']\n",
    "name = filtered_df['ID'] \n",
    "species_dict = {}\n",
    "\n",
    "for i in range(10):\n",
    "    ID = name.iloc[i]\n",
    "    spec = taxonomy.iloc[i]\n",
    "    parts = spec.split(\";\")\n",
    "    for word in parts:\n",
    "        if word in eukarya_classes:\n",
    "            taxo_dict[ID] = word\n",
    "#taxo_dict\n",
    "#len(taxo_dict.keys())\n",
    "\n",
    "# Make a df and ID column is set as index\n",
    "taxo_info = pd.DataFrame.from_dict(taxo_dict, orient='index') \n",
    "L1_taxo = taxo_info.reset_index() \n",
    "L1_taxo.columns = ['ID', 'Taxonomy']\n",
    "L1_taxo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "cF9NSvQDpfde"
   },
   "source": [
    "# Retrieving count number (1357 out of 26431 are valid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "id": "X-BSIh9AyikC"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>All</th>\n",
       "      <th>Non_redundant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1M5_orf2DF0000008</td>\n",
       "      <td>337941</td>\n",
       "      <td>114873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1M1_5endDF0000227</td>\n",
       "      <td>96690</td>\n",
       "      <td>11185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1M1_orf2DF0000228</td>\n",
       "      <td>277575</td>\n",
       "      <td>11129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1M2_orf2DF0000236</td>\n",
       "      <td>296693</td>\n",
       "      <td>17500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1M3_orf2DF0000244</td>\n",
       "      <td>299220</td>\n",
       "      <td>21747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1352</th>\n",
       "      <td>TE3262_SO2_FAM455DF0289703</td>\n",
       "      <td>12387</td>\n",
       "      <td>840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1353</th>\n",
       "      <td>TE3263_SO2_FAM455DF0289704</td>\n",
       "      <td>12914</td>\n",
       "      <td>900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1354</th>\n",
       "      <td>TE3264_SO2_FAM465DF0289705</td>\n",
       "      <td>3843</td>\n",
       "      <td>285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>TE3265_SO2_FAM465DF0289706</td>\n",
       "      <td>1472</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1356</th>\n",
       "      <td>TE3269_SO2_FAM497DF0289710</td>\n",
       "      <td>1297</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1357 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ID     All  Non_redundant\n",
       "0             L1M5_orf2DF0000008  337941         114873\n",
       "1             L1M1_5endDF0000227   96690          11185\n",
       "2             L1M1_orf2DF0000228  277575          11129\n",
       "3             L1M2_orf2DF0000236  296693          17500\n",
       "4             L1M3_orf2DF0000244  299220          21747\n",
       "...                          ...     ...            ...\n",
       "1352  TE3262_SO2_FAM455DF0289703   12387            840\n",
       "1353  TE3263_SO2_FAM455DF0289704   12914            900\n",
       "1354  TE3264_SO2_FAM465DF0289705    3843            285\n",
       "1355  TE3265_SO2_FAM465DF0289706    1472            166\n",
       "1356  TE3269_SO2_FAM497DF0289710    1297            166\n",
       "\n",
       "[1357 rows x 3 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select desired columns and create empty output dict\n",
    "name = filtered_df['ID']  \n",
    "counts = filtered_df['Counts']\n",
    "output_dict = {}\n",
    "\n",
    "# -- testing the number of column with NaN values (only 1357 out of 26431 have valid values)\n",
    "#nonNA_counts = filtered_df['Counts'].dropna()\n",
    "#len(nonNA_counts)\n",
    "\n",
    "# iterating over dataset\n",
    "for i in range(len(filtered_df)):\n",
    "    ID = name.iloc[i]\n",
    "    if type(counts.iloc[i]) == str: # must be string to be converted to dict\n",
    "      hmm_dict = eval(counts.iloc[i])\n",
    "      count_list = list(hmm_dict.values())  \n",
    "      count_dict = eval(str(count_list[0])) \n",
    "      all = count_dict['gathering_all']\n",
    "      non_redun = count_dict['gathering_nonredundant']\n",
    "      #print(ID , all, non_redun)\n",
    "      output_dict[ID] = (all,non_redun) # associate both number to each ID\n",
    "\n",
    "# Display output dict\n",
    "#output_dict\n",
    "\n",
    "# Make a df and ID column is set as index\n",
    "count_df = pd.DataFrame.from_dict(output_dict, orient='index') \n",
    "#count_df\n",
    "\n",
    "# Resetting index to displace ID col\n",
    "ORF2p_counts = count_df.reset_index()\n",
    "\n",
    "# Renaming columns\n",
    "ORF2p_counts.columns = ['ID','All', 'Non_redundant']\n",
    "ORF2p_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "KqFKMLTWo-RH"
   },
   "source": [
    "# Protein extraction with simpler ORFinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "id": "IRXCq6r1tAuS"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1M5_orf2DF0000008</td>\n",
       "      <td>MVDLNPXISIITLNVNGLNTPIKRQRLSDWIKKQDPTICCLQETHF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1M1_orf2DF0000228</td>\n",
       "      <td>MAGVSPYLSIITLNVNGLNSPIKRHRVAEWMKKQDPXICCLQETHF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1M2_orf2DF0000236</td>\n",
       "      <td>MAXVSPYLSIITLNVNGLNSPIKRHRXAEWIKKQDPTICCLQETHF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1M3_orf2DF0000244</td>\n",
       "      <td>MAIVSPXLSIITLNVNGLNSPIKRHRVAEWIKKQDPTICCLQETHF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1M4_orf2DF0000250</td>\n",
       "      <td>MADVNPTXSVITLNVNGLNTPIKRQRLAEWIKKHDPTICCLQETHF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>TE3105_SO2_FAM184DF0289547</td>\n",
       "      <td>TRSKVLLQPLFPPQTYPFFLPPMHGISQINLHHAKGSSATIARLFE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>TE3121_SO2_FAM192DF0289563</td>\n",
       "      <td>PLXGGGTSVARGGWEIPCPPYNRGTLTYPRNLMHGISQINLHHAKG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>TE3146_SO2_FAM209DF0289588</td>\n",
       "      <td>RRGFLTTRMAPSQPGLRVVQINLNHCEAATEDLMLFMSEKKVDVAL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>TE3261_SO2_FAM454DF0289702</td>\n",
       "      <td>ISENGLPTAVQKRNGTLTVNHFPKILMADKNDRMALSPGHPLSGER...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>TE3265_SO2_FAM465DF0289706</td>\n",
       "      <td>GCNASSPPRGTPWIASNYRRKLNELAQLATKQSYLDLLVNCITMVQ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>721 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             ID  \\\n",
       "0            L1M5_orf2DF0000008   \n",
       "1            L1M1_orf2DF0000228   \n",
       "2            L1M2_orf2DF0000236   \n",
       "3            L1M3_orf2DF0000244   \n",
       "4            L1M4_orf2DF0000250   \n",
       "..                          ...   \n",
       "716  TE3105_SO2_FAM184DF0289547   \n",
       "717  TE3121_SO2_FAM192DF0289563   \n",
       "718  TE3146_SO2_FAM209DF0289588   \n",
       "719  TE3261_SO2_FAM454DF0289702   \n",
       "720  TE3265_SO2_FAM465DF0289706   \n",
       "\n",
       "                                                   Seq  \n",
       "0    MVDLNPXISIITLNVNGLNTPIKRQRLSDWIKKQDPTICCLQETHF...  \n",
       "1    MAGVSPYLSIITLNVNGLNSPIKRHRVAEWMKKQDPXICCLQETHF...  \n",
       "2    MAXVSPYLSIITLNVNGLNSPIKRHRXAEWIKKQDPTICCLQETHF...  \n",
       "3    MAIVSPXLSIITLNVNGLNSPIKRHRVAEWIKKQDPTICCLQETHF...  \n",
       "4    MADVNPTXSVITLNVNGLNTPIKRQRLAEWIKKHDPTICCLQETHF...  \n",
       "..                                                 ...  \n",
       "716  TRSKVLLQPLFPPQTYPFFLPPMHGISQINLHHAKGSSATIARLFE...  \n",
       "717  PLXGGGTSVARGGWEIPCPPYNRGTLTYPRNLMHGISQINLHHAKG...  \n",
       "718  RRGFLTTRMAPSQPGLRVVQINLNHCEAATEDLMLFMSEKKVDVAL...  \n",
       "719  ISENGLPTAVQKRNGTLTVNHFPKILMADKNDRMALSPGHPLSGER...  \n",
       "720  GCNASSPPRGTPWIASNYRRKLNELAQLATKQSYLDLLVNCITMVQ...  \n",
       "\n",
       "[721 rows x 2 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -- Extracting all protein sequences from DNA with simpler finder\n",
    "prot_list = []\n",
    "prot_dict = {}\n",
    "name = filtered_df['ID']  \n",
    "DNA = filtered_df['DNA_seq'] \n",
    "\n",
    "for i in range(len(filtered_df)):    #There's no NaN values in DNA column\n",
    "  ID = name.iloc[i]\n",
    "  if type(DNA.iloc[i]) == str:\n",
    "    record = Seq(DNA.iloc[i])\n",
    "  for strand, nuc in [(+1, record), (-1, record.reverse_complement())]:   # explore seq and rev complement\n",
    "       for frame in range(3): #explore all 3 ORFs\n",
    "         min_pro_len = 1000\n",
    "         table = 1\n",
    "         length = 3 * ((len(record) - frame) // 3)  # Multiple of three\n",
    "         for pro in nuc[frame : frame + length].translate(table).split(\"*\"):\n",
    "             if len(pro) >= min_pro_len:\n",
    "               #print( \"%s - %s...%s - length %i,\" % (ID, pro[:20], pro[-3:], len(pro)))\n",
    "               prot_list.append(str(pro))\n",
    "               prot_dict[ID] = str(pro)\n",
    "\n",
    "# Display output prot dict\n",
    "#prot_dict\n",
    "\n",
    "# Make a df and ID column is set as index\n",
    "prot_df = pd.DataFrame.from_dict(prot_dict, orient='index') \n",
    "# Resetting index to displace ID col\n",
    "proteins_df = prot_df.reset_index()\n",
    "\n",
    "# Using drop() function to delete last row and renaming cols\n",
    "proteins_df.columns = ['ID','Seq']\n",
    "#final_df.drop(index=final_df.index[-1], inplace=True)\n",
    "proteins_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "TKRaEFDDpIqf"
   },
   "source": [
    "# Exporting as fasta file from prot_dict{}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "KSkzCj5ouWYx"
   },
   "outputs": [],
   "source": [
    "# output file name\n",
    "output_file = 'ORF2p_v2.fasta'\n",
    "\n",
    "# open output file for writing\n",
    "with open(output_file, 'w') as out_file:\n",
    "  for seq_id, seq in prot_dict.items(): # loop through dictionary items \n",
    "    seq_record = SeqIO.SeqRecord(Seq(seq), id= str(seq_id), description=\"\") # make sure to save ID as string\n",
    "    SeqIO.write(seq_record, out_file, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "aRJVVVbA2uVr"
   },
   "source": [
    "## Merging ORF2p seqs and count number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "id": "JIbhePNL7bOi"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Seq</th>\n",
       "      <th>All</th>\n",
       "      <th>Non_redundant</th>\n",
       "      <th>Taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1M5_orf2DF0000008</td>\n",
       "      <td>MVDLNPXISIITLNVNGLNTPIKRQRLSDWIKKQDPTICCLQETHF...</td>\n",
       "      <td>337941</td>\n",
       "      <td>114873</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1M1_orf2DF0000228</td>\n",
       "      <td>MAGVSPYLSIITLNVNGLNSPIKRHRVAEWMKKQDPXICCLQETHF...</td>\n",
       "      <td>277575</td>\n",
       "      <td>11129</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1M2_orf2DF0000236</td>\n",
       "      <td>MAXVSPYLSIITLNVNGLNSPIKRHRXAEWIKKQDPTICCLQETHF...</td>\n",
       "      <td>296693</td>\n",
       "      <td>17500</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1M3_orf2DF0000244</td>\n",
       "      <td>MAIVSPXLSIITLNVNGLNSPIKRHRVAEWIKKQDPTICCLQETHF...</td>\n",
       "      <td>299220</td>\n",
       "      <td>21747</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1M4_orf2DF0000250</td>\n",
       "      <td>MADVNPTXSVITLNVNGLNTPIKRQRLAEWIKKHDPTICCLQETHF...</td>\n",
       "      <td>313178</td>\n",
       "      <td>32089</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>TE3105_SO2_FAM184DF0289547</td>\n",
       "      <td>TRSKVLLQPLFPPQTYPFFLPPMHGISQINLHHAKGSSATIARLFE...</td>\n",
       "      <td>50037</td>\n",
       "      <td>2731</td>\n",
       "      <td>Insecta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>TE3121_SO2_FAM192DF0289563</td>\n",
       "      <td>PLXGGGTSVARGGWEIPCPPYNRGTLTYPRNLMHGISQINLHHAKG...</td>\n",
       "      <td>50229</td>\n",
       "      <td>1546</td>\n",
       "      <td>Insecta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>TE3146_SO2_FAM209DF0289588</td>\n",
       "      <td>RRGFLTTRMAPSQPGLRVVQINLNHCEAATEDLMLFMSEKKVDVAL...</td>\n",
       "      <td>54467</td>\n",
       "      <td>2789</td>\n",
       "      <td>Insecta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>TE3261_SO2_FAM454DF0289702</td>\n",
       "      <td>ISENGLPTAVQKRNGTLTVNHFPKILMADKNDRMALSPGHPLSGER...</td>\n",
       "      <td>9477</td>\n",
       "      <td>796</td>\n",
       "      <td>Insecta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>TE3265_SO2_FAM465DF0289706</td>\n",
       "      <td>GCNASSPPRGTPWIASNYRRKLNELAQLATKQSYLDLLVNCITMVQ...</td>\n",
       "      <td>1472</td>\n",
       "      <td>166</td>\n",
       "      <td>Insecta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>720 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             ID  \\\n",
       "0            L1M5_orf2DF0000008   \n",
       "1            L1M1_orf2DF0000228   \n",
       "2            L1M2_orf2DF0000236   \n",
       "3            L1M3_orf2DF0000244   \n",
       "4            L1M4_orf2DF0000250   \n",
       "..                          ...   \n",
       "715  TE3105_SO2_FAM184DF0289547   \n",
       "716  TE3121_SO2_FAM192DF0289563   \n",
       "717  TE3146_SO2_FAM209DF0289588   \n",
       "718  TE3261_SO2_FAM454DF0289702   \n",
       "719  TE3265_SO2_FAM465DF0289706   \n",
       "\n",
       "                                                   Seq     All  Non_redundant  \\\n",
       "0    MVDLNPXISIITLNVNGLNTPIKRQRLSDWIKKQDPTICCLQETHF...  337941         114873   \n",
       "1    MAGVSPYLSIITLNVNGLNSPIKRHRVAEWMKKQDPXICCLQETHF...  277575          11129   \n",
       "2    MAXVSPYLSIITLNVNGLNSPIKRHRXAEWIKKQDPTICCLQETHF...  296693          17500   \n",
       "3    MAIVSPXLSIITLNVNGLNSPIKRHRVAEWIKKQDPTICCLQETHF...  299220          21747   \n",
       "4    MADVNPTXSVITLNVNGLNTPIKRQRLAEWIKKHDPTICCLQETHF...  313178          32089   \n",
       "..                                                 ...     ...            ...   \n",
       "715  TRSKVLLQPLFPPQTYPFFLPPMHGISQINLHHAKGSSATIARLFE...   50037           2731   \n",
       "716  PLXGGGTSVARGGWEIPCPPYNRGTLTYPRNLMHGISQINLHHAKG...   50229           1546   \n",
       "717  RRGFLTTRMAPSQPGLRVVQINLNHCEAATEDLMLFMSEKKVDVAL...   54467           2789   \n",
       "718  ISENGLPTAVQKRNGTLTVNHFPKILMADKNDRMALSPGHPLSGER...    9477            796   \n",
       "719  GCNASSPPRGTPWIASNYRRKLNELAQLATKQSYLDLLVNCITMVQ...    1472            166   \n",
       "\n",
       "     Taxonomy  \n",
       "0    Mammalia  \n",
       "1    Mammalia  \n",
       "2    Mammalia  \n",
       "3    Mammalia  \n",
       "4    Mammalia  \n",
       "..        ...  \n",
       "715   Insecta  \n",
       "716   Insecta  \n",
       "717   Insecta  \n",
       "718   Insecta  \n",
       "719   Insecta  \n",
       "\n",
       "[720 rows x 5 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -- Merging both datasets and exporting as csv\n",
    "\n",
    "ORF2p_seqs_counts = pd.merge(proteins_df,ORF2p_counts)\n",
    "ORF2p_with_taxo = pd.merge(ORF2p_seqs_counts, L1_taxo)\n",
    "\n",
    "ORF2p_with_taxo\n",
    "\n",
    "#output_path = \"/Users/leandrojorqueravalero/Desktop/PhD/Synthetic_ORF2p\"\n",
    "\n",
    "#ORF2p_with_taxo.to_csv(r'/Users/leandrojorqueravalero/Desktop/PhD/synthetic_ORF2/data/ORF2p_taxo_v2.csv', index= False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "AUeE8Ytjozl5"
   },
   "source": [
    "# Protein extraction with ORFinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "RE7JiPB6gpTv"
   },
   "outputs": [],
   "source": [
    "#record = SeqIO.read(\"NC_005816.fna\", \"fasta\") #parse sequence\n",
    "table = 1  #set table for aa\n",
    "min_pro_len = 1000 #lower limit for protein length\n",
    "\n",
    "def find_orfs_with_trans(seq, trans_table,min_protein_lenght):\n",
    "  answer=[] #list of results\n",
    "  seq_len = len(seq)\n",
    "  for strand, nuc in ([+1, seq], (-1, seq.reverse_complement())):\n",
    "    for frame in range(3):\n",
    "      trans = nuc[frame:].translate(trans_table)\n",
    "      trans_len = len(trans)\n",
    "      aa_start = 0\n",
    "      aa_end = 0\n",
    "      while aa_start < trans_len:\n",
    "        aa_end = trans.find(\"*\",aa_start)\n",
    "        if aa_end == -1:\n",
    "          aa_end = trans_len\n",
    "        if aa_end - aa_start >= min_protein_lenght:\n",
    "          if aa_end == 1:\n",
    "            start = frame + aa_start * 3\n",
    "            end = min(seq_len, frame + aa_end * 3 +3)\n",
    "          else:\n",
    "            start = seq_len - frame - aa_end * 3 -3\n",
    "            end = seq_len - frame - aa_start * 3\n",
    "          answer.append ((start, end, strand, trans[aa_start:aa_end]))\n",
    "        aa_start = aa_end + 1 \n",
    "\n",
    "  answer.sort()  #sort the table for start base position\n",
    "  return answer\n",
    "\n",
    "# -- iterate over all elements in DNA_seq col\n",
    "record_list=[]\n",
    "for i in filtered_df['DNA_seq']:\n",
    "  if type(i) == str:\n",
    "    record = Seq(i)\n",
    "    record_list.append(record)\n",
    "    #print (record)\n",
    "    orf_list = find_orfs_with_trans(record, table, min_pro_len) # -- call ORFinder\n",
    "  else:\n",
    "    continue\n",
    "#len(record_list)\n",
    "#len(orf_list)\n",
    "for start, end, strand, pro in orf_list:  # -- show results \n",
    "  print (\"%s ...%s - lenght %i, strand %i, %i:%i\" % (pro[:20], pro[-3:], len(pro), strand, start, end))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "3e2DSdhwO0-M"
   },
   "outputs": [],
   "source": [
    "# -- iterate over all elements in DNA_seq col using complex ORFinder\n",
    "record_list=[]\n",
    "for i in filtered_df['DNA_seq']:\n",
    "  if type(i) == str:\n",
    "    record = Seq(i)\n",
    "    record_list.append(record)\n",
    "    #print (record)\n",
    "    orf_list = find_orfs_with_trans(record, table, min_pro_len) # -- call ORFinder\n",
    "  else:\n",
    "    continue\n",
    "#len(record_list)\n",
    "#len(orf_list)\n",
    "for start, end, strand, pro in orf_list:  # -- show results \n",
    "  print (\"%s ...%s - lenght %i, strand %i, %i:%i\" % (pro[:20], pro[-3:], len(pro), strand, start, end))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "cZOBRtL_GQqB"
   },
   "source": [
    "# Table supplementary ORF2/hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 411
    },
    "id": "a2Vxwa7pGPFH",
    "outputId": "ebfb242a-2fd6-4c50-8daa-5aa428656b22"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-a352c507-02d1-499b-ae1e-cf5f6819d6f9\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>ORF2_CN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>1607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Balaenoptera\\nacutorostrata scammoni</td>\n",
       "      <td>1338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Rhinopithecus roxellana</td>\n",
       "      <td>714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>Brassica napus</td>\n",
       "      <td>544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Schistosoma curassoni</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Schistosoma haematobium</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>Schistosoma japonicum</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>Schistosoma mansoni</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>Orussus abietinus</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>407 rows × 2 columns</p>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a352c507-02d1-499b-ae1e-cf5f6819d6f9')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-a352c507-02d1-499b-ae1e-cf5f6819d6f9 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-a352c507-02d1-499b-ae1e-cf5f6819d6f9');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "                                  species  ORF2_CN\n",
       "0                   Monodelphis domestica     1607\n",
       "44   Balaenoptera\\nacutorostrata scammoni     1338\n",
       "84                Rhinopithecus roxellana      714\n",
       "75                           Mus musculus      629\n",
       "350                        Brassica napus      544\n",
       "..                                    ...      ...\n",
       "262                 Schistosoma curassoni        0\n",
       "263               Schistosoma haematobium        0\n",
       "264                 Schistosoma japonicum        0\n",
       "265                   Schistosoma mansoni        0\n",
       "246                     Orussus abietinus        0\n",
       "\n",
       "[407 rows x 2 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df_2 = pd.read_csv(\"/content/ORF2_suppl.csv\", delimiter=';')\n",
    "df_2.columns = ['species', 'ORF2_CN']\n",
    "#print(df)\n",
    "#df.head()\n",
    "df_2.sort_values(by=['ORF2_CN'], ascending=False)\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [
    "BNG40Ig6PX1L",
    "cZOBRtL_GQqB"
   ],
   "include_colab_link": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
